### PR TITLE
Move ObjectFromHandle to the GCHandleTable interface

### PR DESCRIPTION
**Pull request by adityamandaleeka**
This change touches a lot of files, but it's essentially the same idea everywhere: Rather than using the `ObjectFromHandle` macro that was in objecthandle.h, put `ObjectFromHandle` on the `GCHandleTable` interface and modify the callers to call that. 

The DAC will use HndFetchHandle directly since for now we don't want to deal with getting it to call virtual functions on the GC side.

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@Maoni0 @swgillespie @sergiy-k PTAL

**Comment by jkotas** [comment]
It is ok to do the debug-only validation via the interface, but it has to stay as simple pointer dereference in the release builds.

**Comment by adityamandaleeka** [comment]
You're right. I'll make an `ObjectFromHandle` on the VM side that calls the interface method in debug and does a simple dereference in release.

**Comment by adityamandaleeka** [comment]
@jkotas @Maoni0 PTAL

File: src/gc/gchandletable.cpp [hunk 1]
```diff
@@ -22,3 +22,8 @@ void GCHandleTable::Shutdown()
 {
     Ref_Shutdown();
 }
+
+Object* GCHandleTable::ObjectFromHandle(OBJECTHANDLE handle)
+{
+    return OBJECTREFToObject(::HndFetchHandle(handle));
```
Comment (jkotas) [inline src/gc/gchandletable.cpp]: When you look at what HndFetchHandle actually does, it gets AppDomain index, and then validates a bunch of things on the AppDomain. All these are bad dependencies for local GC. Have you thought about what we want to about those?

I am thinking that the whole validation should stay on the VM side, and the only method on the GC/EE interface should be a method to fetch AppDomain index from handle table. (It would be nice to not call it AppDomain index, but something more general - like `void* context`.)

Comment (Maoni0) [inline src/gc/gchandletable.cpp]: yep, @swgillespie and I talked about what should be done wrt AD. aside from the alloc/survived bytes info that GC tracks for appdomain resource manager purpose, ideally everything else should goto the EE side.

Comment (adityamandaleeka) [inline src/gc/gchandletable.cpp]: OK. I was already planning to expose the AppDomain index for a handle table via the interface, so I'll do what @jkotas suggested above.

File: src/gc/gchandletable.cpp [hunk 2]
```diff
@@ -22,3 +22,13 @@ void GCHandleTable::Shutdown()
 {
     Ref_Shutdown();
 }
+
+uint32_t GCHandleTable::GetHandleTableContext(HHANDLETABLE hTable)
```
Comment (jkotas) [inline src/gc/gchandletable.cpp]: I would make this `void*` and cast it to int32_t on the VM side.

File: src/gc/gcinterface.h [hunk 3]
```diff
@@ -149,10 +149,18 @@ struct segment_info
...
 };
 
+
+/*
+ * handle to handle table
+ */
+typedef DPTR(struct HandleTable) PTR_HandleTable;
+typedef DPTR(PTR_HandleTable) PTR_PTR_HandleTable;
+typedef PTR_HandleTable HHANDLETABLE;
+typedef PTR_PTR_HandleTable PTR_HHANDLETABLE;
```
Comment (jkotas) [inline src/gc/gcinterface.h]: This leaks the impl details to the interface. I think HHANDLETABLE should be void* on the interface, and it should be case internally to PTR_HandleTable, etc.

File: src/vm/exceptionhandling.h [hunk 4]
```diff
@@ -253,7 +253,11 @@ class ExceptionTracker
 
         if (NULL != m_hThrowable)
         {
+#ifndef DACCESS_COMPILE
             return ObjectFromHandle(m_hThrowable);
```
Comment (jkotas) [inline src/vm/exceptionhandling.h]: Is this change still needed?

**Comment by jkotas** [comment]
LGTM otherwise

**Comment by adityamandaleeka** [comment]
@jkotas Thanks, I've addressed your feedback.

File: src/gc/gchandletable.cpp [hunk 5]
```diff
@@ -22,3 +22,13 @@ void GCHandleTable::Shutdown()
 {
     Ref_Shutdown();
 }
+
+void* GCHandleTable::GetHandleTableContext(void* handleTable)
+{
+    return (void*)((uintptr_t)::HndGetHandleTableADIndex((HHANDLETABLE)handleTable).m_dwIndex);
```
Comment (jkotas) [inline src/gc/gchandletable.cpp]: In future PR, we should also change `m_dwIndex` to `void* m_pContext`.
